Library/Formula/mlton.rb

```diff
@@ -0,0 +1,131 @@
+  # Replace these with placeholders for the Homebrew GMP location.
+  patch :DATA
+
+  def install
+    # Finish configuring source release with Homebrew GMP location.
+    inreplace "bin/mlton-script", "gmpHomebrewCCOpts=''", "gmpHomebrewCCOpts='-I#{Formula["gmp"].opt_prefix}/include'"
+    inreplace "bin/mlton-script", "gmpHomebrewLinkOpts=''", "gmpHomebrewLinkOpts='-L#{Formula["gmp"].opt_prefix}/lib'"
+    inreplace "runtime/Makefile", "GMP_HOMEBREW_XCFLAGS :=", "GMP_HOMEBREW_XCFLAGS := -I#{Formula["gmp"].opt_prefix}/include"
+
+    # Import the corresponding upstream binary release to 'bootstrap'.
+    bootstrap = buildpath/"bootstrap"
+    bootstrap_bin = bootstrap/"usr/local/bin"
+    bootstrap_lib = bootstrap/"usr/local/lib"
+    mkdir bootstrap
+    resource("bootstrap").stage(bootstrap/"usr")
+    # The binary release assumes that the MLton support files are in
+    # '/usr/local/lib/mlton' and that GMP is in '/usr/local',
+    # '/opt/local' (MacPorts), or '/sw' (Fink).
+    # Would like to patch/inreplace the
+    # 'bootstrap/usr/local/bin/mlton' script as is done above for the
+    # source release, but the Homebrew patch API does not support
+    # patching external resources.  Since the binary and source
+    # releases correspond, the 'bootstrap/usr/local/bin/mlton' script
+    # is identical to the (original, un-patched/un-inreplaced)
+    # 'bin/mlton-script', modulo a 'lib=...' that determines the
+    # location of the MLton support files.  Therefore, use the
+    # (patched/inreplaced) 'bin/mlton-script' for the
+    # 'bootstrap/usr/local/bin/mlton' script, updating 'lib=...' to
+    # the location of the bootstrap MLton support files.
+    cp "bin/mlton-script", bootstrap_bin/"mlton"
+    inreplace bootstrap_bin/"mlton", /^lib=.*/, "lib='#{bootstrap_lib}/mlton'"
+    chmod "a+x", bootstrap_bin/"mlton"
+    ENV.prepend_path "PATH", bootstrap_bin
+
+    ENV.deparallelize
+    system "make", "all-no-docs"
+    system "make", "install-no-docs", "DESTDIR=", "PREFIX=#{prefix}", "MAN_PREFIX_EXTRA=/share"
+  end
+
+  test do
+    (testpath/"hello.sml").write <<-'EOS'.undent
+      val () = print "Hello, Homebrew!\n"
+    EOS
+    system "#{bin}/mlton", "hello.sml"
+    assert_equal "Hello, Homebrew!\n", `./hello`
+  end
+end
+
+__END__
+diff --git a/bin/mlton-script b/bin/mlton-script
+index ed4cc38..3af01dd 100644
+--- a/bin/mlton-script
++++ b/bin/mlton-script
+@@ -80,18 +80,8 @@ doit () {
+ # You may need to add a line with -link-opt '-L/path/to/libgmp' so
+ # that the linker can find libgmp.
+ 
+-# The darwin linker complains (loudly) about non-existent library
+-# search paths.
+-darwinLinkOpts=''
+-if [ -d '/usr/local/lib' ]; then
+-        darwinLinkOpts="$darwinLinkOpts -L/usr/local/lib"
+-fi
+-if [ -d '/opt/local/lib' ]; then
+-        darwinLinkOpts="$darwinLinkOpts -L/opt/local/lib"
+-fi
+-if [ -d '/sw/lib' ]; then
+-        darwinLinkOpts="$darwinLinkOpts -L/sw/lib"
+-fi
++gmpHomebrewCCOpts=''
```

Comment:
If we are going to patch it like this, I think we should probably leave all the existing `$darwinLinkOpts` code in place, and just add another line to get the Homebrew prefix (determined at install time) in there and in a `$darwinCcOpts`. That will minimize the difference in behavior between the Homebrew-installed and regular MLtons.